habitat/tasks/rearrange/rearrange_task.py

```diff
@@ -203,3 +205,24 @@ def should_end(self, new_val: bool):
                 f"-----Episode {self._episode_id} requested to end after {self._cur_episode_step} steps.-----"
             )
             rearrange_logger.debug("-" * 40)
+
+    def get_metrics_at_episode_end(self):
+
+        metrics: Dict[str, Any] = {}
+
+        # Add a gfx-replay list of keyframes for the episode. This is a JSON string that
+        # should be saved to a file; the file can be read by visualization tools
+        # (e.g. import into Blender for screenshots and videos).
+        if (
+            self._episode_id
+            and self._sim.sim_config.sim_cfg.enable_gfx_replay_save
+        ):
+            if not self._gfx_replay_keyframes_string:
+                self._gfx_replay_keyframes_string = (
```

Comment:
Note we cache the result of `write_saved_keyframes_to_string()` instead of allowing it to perhaps be called twice for the same episode (which wouldn't work correctly).